### PR TITLE
Add composition root tests

### PR DESCRIPTION
**Pull request by Sondergaard**
## Description

The proposed change will automatically discover when new Functions or Web Api controllers are added to the solution.

This will catch the exceptions where a new function or controller is added, but a test case is not created to test for the needed dependencies.

With the use of reflection constructor dependencies are located and will then be tested against the service collection for the host.

The result is shown in the test explorer:
image

### CODE REVIEW

**Comment by Sondergaard** [comment]
Grouped related tests - the test output is now:
image

File: source/dotnet/Services/IntegrationTests/CompositionRootTests.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+using Microsoft.AspNetCore.Mvc;
+using Xunit;
+using iel = Energinet.DataHub.Wholesale.IntegrationEventListener;
+using pm = Energinet.DataHub.Wholesale.ProcessManager;
+using snd = Energinet.DataHub.Wholesale.Sender;
+using wapi = Energinet.DataHub.Wholesale.WebApi;
+
+namespace Energinet.DataHub.Wholesale.IntegrationTests;
+
+[Collection(nameof(SenderIntegrationTestHost))]
```
Comment (BjarkeMeier) [inline source/dotnet/Services/IntegrationTests/CompositionRootTests.cs:26]: Is this a proper collection name?

File: source/dotnet/Services/IntegrationTests/CompositionRootTests.cs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        var constructorDependencies = ReflectionHelper.FindAllConstructorDependenciesForType();
+
+        return typeof(wapi.Program).Assembly.GetTypes()
+            .Where(t => t.IsSubclassOf(typeof(ControllerBase)))
+            .Select(t => new object[] { new Requirement(t.Name, constructorDependencies(t), t) });
+    }
+
+    #endregion
+
+    [Theory(DisplayName = nameof(SenderIntegrationFunctions_can_resolve_dependencies_for))]
```
Comment (BjarkeMeier) [inline source/dotnet/Services/IntegrationTests/CompositionRootTests.cs:51]: What is the purpose of these display names when they are identical to the method names?

File: source/dotnet/Services/IntegrationTests/TestHelpers/ReflectionHelper.cs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Reflection;
+using Microsoft.Azure.Functions.Worker;
+
+namespace Energinet.DataHub.Wholesale.IntegrationTests.TestHelpers;
+
+public static class ReflectionHelper
```
Comment (BjarkeMeier) [inline source/dotnet/Services/IntegrationTests/TestHelpers/ReflectionHelper.cs:20]: I was confused about the method names. Expected them to return types. Can we use names that indicate that they return functions/providers?

File: source/dotnet/Services/IntegrationTests/TestHelpers/Requirement.cs [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Energinet.DataHub.Wholesale.IntegrationTests.TestHelpers;
+
+public record Requirement(string Name, IEnumerable<Type> DependentOn, Type? ActualType = null)
```
Comment (BjarkeMeier) [inline source/dotnet/Services/IntegrationTests/TestHelpers/Requirement.cs:19]: Can we add documentation about what this is and the purpose of it?

**Review by BjarkeMeier — APPROVED** [review]
Nice with the generic reflection-based approach :-)

File: source/dotnet/Services/IntegrationTests/CompositionRootTests.cs (hunk 1 above)
Comment (Sondergaard) [reply source/dotnet/Services/IntegrationTests/CompositionRootTests.cs]: Collection name changed to match that of the class

File: source/dotnet/Services/IntegrationTests/CompositionRootTests.cs (hunk 2 above)
Comment (Sondergaard) [reply source/dotnet/Services/IntegrationTests/CompositionRootTests.cs]: `DisplayName` is removed from the tests

File: source/dotnet/Services/IntegrationTests/TestHelpers/ReflectionHelper.cs (hunk 3 above)
Comment (Sondergaard) [reply source/dotnet/Services/IntegrationTests/TestHelpers/ReflectionHelper.cs]: Class renamed and updated with docs.

File: source/dotnet/Services/IntegrationTests/TestHelpers/Requirement.cs (hunk 4 above)
Comment (Sondergaard) [reply source/dotnet/Services/IntegrationTests/TestHelpers/Requirement.cs]: Documentation added.

**Pull request merged**